src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Telemetry/VsTelemetryService.cs

```diff
@@ -0,0 +1,238 @@
+        /// <param name="description">Description to include with the NFW</param>
+        /// <param name="exception">Exception that caused the NFW</param>
+        /// <param name="callback">Gathers information for the NFW</param>
+        /// <returns>Posted telemetry event.</returns>
+        public TelemetryEventCorrelation Report(string eventPostfix, string description, Exception exception, Func<IFaultUtility, int> callback = null)
+        {
+            return TelemetryService.DefaultSession.PostFault(
+                eventName: $"{EventPrefix}{eventPostfix}",
+                description: description,
+                exceptionObject: exception,
+                gatherEventDetails: callback);
+        }
+
+        /// <summary>
+        /// Posts an event with a single property.
+        /// </summary>
+        /// <param name="eventName">Name of the event</param>
+        /// <param name="propertyName">Name of the property.</param>
+        /// <param name="propertyValue">Value of the property.</param>
+        /// <param name="unconfiguredProject">Correlated project asset for this event.</param>
+        public void PostProperty(string eventName, string propertyName, string propertyValue, UnconfiguredProject unconfiguredProject = null)
+        {
+            TelemetryEvent telemetryEvent = new TelemetryEvent(EventPrefix + eventName.ToLower());
+            telemetryEvent.Properties.Add(BuildPropertyName(eventName, propertyName), propertyValue);
+            TryCorrelateProject(unconfiguredProject, telemetryEvent);
+            TelemetryService.DefaultSession.PostEvent(telemetryEvent);
+        }
+
+        /// <summary>
+        /// Posts an event with multiple properties.
+        /// </summary>
+        /// <param name="eventName">Name of the event.</param>
+        /// <param name="properties">Collection of property name and values.</param>
+        /// <param name="unconfiguredProject">Correlated project asset for this event.</param>
+        public void PostProperties(string eventName, IEnumerable<(string propertyName, string propertyValue)> properties, UnconfiguredProject unconfiguredProject = null)
+        {
+            TelemetryEvent telemetryEvent = new TelemetryEvent(EventPrefix + eventName.ToLower());
+            foreach(var property in properties)
+            {
+                telemetryEvent.Properties.Add(BuildPropertyName(eventName, property.propertyName), property.propertyValue);
+            }
+
+            TryCorrelateProject(unconfiguredProject, telemetryEvent);
+            TelemetryService.DefaultSession.PostEvent(telemetryEvent);
+        }
+
+        /// <summary>
+        /// Hashes personally identifiable information for telemetry consumption.
+        /// </summary>
+        /// <param name="value">Value to hashed.</param>
+        /// <returns>Hashed value.</returns>
+        public string HashValue(string value)
+        {
+            using (var md5 = MD5.Create())
```

Comment:
I wonder if someone was louder on our side... we have about 3-4 instances of this on the C++ side. But either way, I'll make the change to avoid having to go through the process.